virtual-host-gatherer/lib/gatherer/modules/Xen.py

```diff
@@ -0,0 +1,169 @@
+from gatherer.modules import WorkerInterface
+
+try:
+    import XenAPI
+    IS_VALID = True
+except ImportError:
+    IS_VALID = False
+
+
+
+class Xen(WorkerInterface):
+    """
+    Worker class for Xen.
+    """
+
+    DEFAULT_PARAMETERS = OrderedDict([
+        ('hostname', ''),
+        ('port', '80'),
+        # ('port', '443'),
+        ('username', ''),
+        ('password', '')])
+
+    VMSTATE = {
+        'Halted': 'stopped',
+        'Running': 'running',
+        'Paused': 'paused',
+        'Suspended': 'suspended',
+    }
+
+    def __init__(self):
+        """
+        Constructor.
+
+        :return:
+        """
+
+        self.log = logging.getLogger(__name__)
+        self.host = self.port = self.user = self.password = None
+
+    # pylint: disable=R0801
+    def set_node(self, node):
+        """
+        Set node information
+
+        :param node: Dictionary of the node description.
+        :return: void
+        """
+
+        try:
+            self._validate_parameters(node)
+        except AttributeError as error:
+            self.log.error(error)
+            raise error
+
+        self.host = node['hostname']
+        self.port = node.get('port', 80)
+        self.user = node['username']
+        self.password = node['password']
+
+    def parameters(self):
+        """
+        Return default parameters
+
+        :return: default parameter dictionary
+        """
+
+        return self.DEFAULT_PARAMETERS
+
+    def run(self):
+        """
+        Start worker.
+
+        :return: Dictionary of the hosts in the worker scope.
+        """
+        output = dict()
+        self.log.info('Connect Xen Server %s:%s as user %s', self.host, self.port, self.user)
+        try:
+            session = XenAPI.Session("http://%s:%s" % (self.host, self.port))
+            session.xenapi.login_with_password(self.user, self.password, "2.3", "Xen Gatherer")
+
+            # Dictionary {opaque_reference: host_data}
+            hosts = {ref: session.xenapi.host.get_record(ref) for ref in session.xenapi.host.get_all()}
+            for _ref, host in hosts.items():
+                self.log.debug('Host=%s, uuid=%s', host["name_label"], host["uuid"])
+
+            vms = []
+            for vm_ref in session.xenapi.VM.get_all():
+                vm = session.xenapi.VM.get_record(vm_ref)
+                if not vm['is_a_template']:
+                    vms.append(vm)
+
+            for ref, host in hosts.items():
+                output[host['name_label']] = {
+                    'name': host['name_label'],
+                    'hostIdentifier': host['name_label'],
+                    'type': 'xen',
+                    'os': 'Xen AHV',
+                    'osVersion': host['software_version']['product_version'],
+                    'totalCpuSockets': host['cpu_info']['socket_count'],
+                    'totalCpuCores': host['cpu_info']['cpu_count'],
+                    'totalCpuThreads': 1,
```

Comment:
Just a comment here: I think we can get the actual value for `totalCpuThreads` from the XenAPI. Maybe you can take some inspiration from the Salt `xapi_virt` module, where something like this is also done: https://github.com/openSUSE/salt/blob/openSUSE-3002.2/salt/modules/xapi_virt.py#L319

Reply:
@meaksh I cannot find in the response a field for this:
```
  "cpu_info": {
    "cpu_count": "4",
    "socket_count": "4",
    "vendor": "GenuineIntel",
    "speed": "2500.014",
    "modelname": "Intel Xeon E312xx (Sandy Bridge, IBRS update)",
    "family": "6",
    "model": "42",
    "stepping": "1",
    "flags": "fpu de tsc msr pae mce cx8 apic sep mca cmov pat clflush mmx fxsr sse sse2 ss syscall nx rdtscp lm constant_tsc rep_good nopl cpuid pni pclmulqdq ssse3 cx16 sse4_1 sse4_2 popcnt aes xsave avx hypervisor lahf_lm cpuid_fault ssbd ibrs ibpb stibp xsaveopt",
    "features": "9fba2223-0f8bfbff-00000001-2c100800",
    "features_pv": "1f89cbf5-96b82203-2991cbf5-00000003-00000001-00000000-00000000-00000000-00001000-8c000400-00000000-00000000-00000000-00000000-00000000",
    "features_hvm": "1f8bfbff-97ba2223-2d93fbff-00000403-00000001-00000002-00000004-00000000-00001000-8c000400-00000000-00000000-00000000-00000000-00000000",
    "features_hvm_host": "1f8bfbff-97ba2223-2d93fbff-00000403-00000001-00000002-00000004-00000000-00001000-8c000400-00000000-00000000-00000000-00000000-00000000",
    "features_pv_host": "1f89cbf5-96b82203-2991cbf5-00000003-00000001-00000000-00000000-00000000-00001000-8c000400-00000000-00000000-00000000-00000000-00000000"
  },
```

Reply:
Looking at the API doc:
https://xapi-project.github.io/xen-api/classes/host.html
```
(string → string) map cpu_configuration 
The CPU configuration on this host. May contain keys such as "nr_nodes", "sockets_per_node", "cores_per_socket", or "threads_per_core"
```
That 'may' suggests that this attribute is not guaranted to be there.  Exactly what I see in my test machine.
This code https://fossies.org/linux/salt/salt/modules/xapi_virt.py
returns `None` if the field is absent:
```
"cputhreads": _get_val(host_rec, ["cpu_configuration", "threads_per_core"]),
```
I would use the following:
```
host['cpu_configuration'].get('threads_per_core', 1)
```
to be consistent with the default we use (1) at `Kubernetes.py`